src/polyfill/bigint64array.js

```diff
@@ -0,0 +1,82 @@
+    return BigInt(lower) | (BigInt(upper) << BigInt(32));
+}
+
+function bigIntToParts(value) {
+    var lower = Number(BigInt(value) & BigInt(0xffffffff)) | 0;
+    var upper = (Number(BigInt(value) >> BigInt(32)) | 0);
+    return [lower, upper];
+}
+
+function createBigIntArrayShim(partsToBigInt) {
+    function createBigInt64Array(array) {
+        let proxy = new Proxy({
+            slice: function(min, max) {
+                var new_buf = array.slice(min * 2, max *2);
+                return createBigInt64Array(new_buf);
+            },
+            subarray: function(min, max) {
+                var new_buf = array.subarray(min * 2, max *2);
+                return createBigInt64Array(new_buf);
+            },
+            [Symbol.iterator]: function*() {
+                for (var i = 0; i < (array.length)/2; i++) {
+                    yield partsToBigInt(array[2*i], array[2*i+1]);
+                }
+            },
+            buffer : array.buffer,
+            byteLength : array.byteLength,
+            offset : array.byteOffset / 2,
+            copyWithin: function(target, start, end) {
+                array.copyWithin(target*2, start * 2, end*2);
+                return proxy;
+            },
+            set: function(source, targetOffset) {
+                if(2*(source.length + targetOffset) > array.length) {
+                    // This is the Chrome error message
+                    // Firefox: "invalid or out-of-range index"
+                    throw new RangeError("offset is out of bounds");
+                }
+                for (var i = 0; i < array.length; i++) {
+                    var value = source[i];
+                    var pair = bigIntToParts(BigInt(value));
+                    array.set(pair, 2*(targetOffset + i));
+                }
+            }
+            }, {
+            get: function(target, idx, receiver) {
+                if (typeof idx !== "string" || !/^\d+$/.test(idx)) {
+                    return Reflect.get(target, idx, receiver);
+                }
+                var lower = array[idx * 2];
+                var upper = array[idx * 2 + 1];
+                return partsToBigInt(lower, upper);
+            },
+            set: function(target, idx, value, receiver) {
+                if (typeof idx !== "string" || !/^\d+$/.test(idx)) {
+                    return Reflect.set(target, idx, value, receiver);
+                }
+                if (typeof value !== "bigint") {
+                    // Chrome error message, Firefox has no "a" in front if "BigInt".
+                    throw new TypeError(`Cannot convert ${value} to a BigInt`);
+                }
+                var pair = bigIntToParts(value);
+                array.set(pair, 2*idx);
+            }
+        });
+        return proxy;
+    }
+    return createBigInt64Array;
+}
+
+var createBigUint64Array = createBigIntArrayShim(partsToBigIntUnsigned);
+var createBigInt64Array = createBigIntArrayShim(partsToBigIntSigned);
```

Comment:
Right, that would be sensible. I got confused about scopes though. You mean I should define `BigUint64Array` on the global scope? How do I find the global scope? Probably `globalThis` works for every browser with `BigInt`.

Reply:
I mean just define it here like:

```
function BigUint64Array() {
   ...
}
```

So it can be used as JS constructor function in preamble.js? 

The fact the its a partial polyfill worries my a little.. so I could do either way on this.

Reply:
It would not be hard to fill in the other methods, I just didn't want to. Most of them are not useful for wasm-related tasks. I guess `fill` is useful, but I can't see methods `entries`, `every`, `filter`, `find`, `findIndex`, `includes`,  `join`, `keys`, `lastIndexOf`, `map`, `reduce`, `reduceRight`, `reverse`, `some`, `sort`, `values`, `toLocaleString`, `toString` as ever being useful.

Reply:
I don't think we need a full polyfill.. i'm just wondering, given that its no a full polyfill if we should actually define at the global scope or not.   I think maybe its fine as long as folks can do "-sNO_POLYFILLS" if they don't want it.

Reply:
I think it's probably better not to define it in the global scope since it's not really necessary and it isn't a full polyfill. But otherwise there are scope issues. I guess we can do:
```js
var BigUint64ArrayShim;
if(typeof BigUint64Array !== "undefined"){
    BigUint64ArrayShim = BigUint64Array;
} else {
    BigUint64ArrayShim = Shim;
}
HEAPU64 = new BigUint64ArrayShim(buf);
```

Reply:
I think it's also better not to gate it behind `POLYFILL` because it's already gated behind `WASM_BIGINT` and `MIN_SAFARI_VERSION`, but either way is fine.